### PR TITLE
Add Fusion variations

### PR DESCRIPTION
**Pull request by hhorii**
### Summary

Add fusion variations

### Details and comments

Currently cost-based-fusion is an only method of fusion. This PR adds 1- and 2-Qubit fusion, which greedily fuse operations with consideration of commutativity in gate operations. To add them, this PR introduces `AER::Transpile::Fuser`, which is the base class of three fusion methods.

- [x] Add N-Qubit fusion
- [x] Add tests

### CODE REVIEW

**Comment by hhorii** [comment]
@vvilpas Tests were failed because of a issue in recent MPI optimization. @doichanj is sending a PR to fix this issue.

File: src/transpile/fusion.hpp
Comment (vvilpas) [inline src/transpile/fusion.hpp:730]: Is there a usecase where we wouldn't want to use all these `fusers` but only a few them (selected by conf for example)?

**Review by vvilpas — COMMENTED** [review]
I leave a few comments, while we wait for Doi's PR. Looks like they are many, but a large fraction of them are just to add forgotten `override` keyword.

File: src/transpile/fusion.hpp
Comment (hhorii) [reply src/transpile/fusion.hpp:558]: This is good for`found`, but not for `u1_end`. Consecutive diagonal gates may exists after CX as follows:

```
  //      ┌───┐                                   ┌───┐
  // q_0: ┤ X ├───────────────────────────────────┤ X ├
  //      └─┬─┘┌───┐            ┌──────────┐ ┌───┐└─┬─┘
  // q_1: ──■──┤ X ├────────────┤ diagonal ├─┤ X ├──■──
  //           └─┬─┘┌──────────┐└──────────┘ └─┬─┘
  // q_2: ───────■──┤ diagonal ├───────────────■───────
  //                └──────────┘
  //        ■ [from]                           ■ [pos,u1_end]
  //             ■ [cx_end]
```

I will add more comments on this code.

File: src/transpile/fusion.hpp
Comment (hhorii) [reply src/transpile/fusion.hpp:614]: True. I will take it. Thanks!

File: src/transpile/fusion.hpp
Comment (hhorii) [reply src/transpile/fusion.hpp:730]: When performance degradation happens because of fusion, we will disable some of them for diagnostics. All of fusers can be disabled with configuration (such as `fusion_enable.diagonal`).

File: src/transpile/fusion.hpp
Comment (vvilpas) [reply src/transpile/fusion.hpp]: Ah, right, you keep increment `pos`. Missed that.

File: src/transpile/fusion.hpp
Comment (vvilpas) [reply src/transpile/fusion.hpp:763]: ok, you always have the fusers and you can disable by config.

**Comment by hhorii** [comment]
#1149 will resolve fails in test.

**Comment by chriseclectic** [comment]
I think this needs rebasing to remove #1149 now that the bug fixes have been merged into master. Otherwise we need to wait for 1149 to be merged before this can be merged

File: src/framework/operations.hpp
Comment (chriseclectic) [inline src/framework/operations.hpp:291]: Since vec is usually a temporary in the fusion code that uses this, you should make a move-semantics version as well with
```suggestion
inline Op make_diagonal(const reg_t &qubits, cvector_t &&vec, std::string label = "") {
  // same but with
  op.params = std::move(vec);

File: src/transpile/fusion.hpp
Comment (chriseclectic) [inline src/transpile/fusion.hpp:43]: Does diagonal mean all ops in fusioned_ops are diagonal, or that the output is diagonal even if the ops are not?
If its the first we should have a specialized simulation method that doesn't require building the full square matrix.

File: src/transpile/fusion.hpp
Comment (chriseclectic) [inline src/transpile/fusion.hpp:68]: To continue from above comment. If all ops are diagonal, we should have `generate_operation_internal` take the diagonal arg and return the diagonal operation if the method supports it.

File: src/transpile/fusion.hpp
Comment (chriseclectic) [inline src/transpile/fusion.hpp:73]: If you add the move semantics to `make_diagonal` this should be 
```suggestion
      fusioned_op = Operations::make_diagonal(fusioned_op.qubits, std::move(vec), std::string("fusion"));
```

File: src/transpile/fusion.hpp
Comment (chriseclectic) [inline src/transpile/fusion.hpp:161]: We could add a diagonal_superop instruction for the density matrix / superop simulators

File: src/transpile/fusion.hpp
Comment (chriseclectic) [inline src/transpile/fusion.hpp:518]: Won't it be expensive to check every matrix is diagonal?

**Review by chriseclectic — COMMENTED** [review]
I have a couple questions about diagonal param in fusion. The main one is if we are fusing a set of diagonal matrices, we should not use the unitary simulator since that is inefficient, but directly update the diagonal vector instead.

File: src/transpile/fusion.hpp
Comment (hhorii) [reply src/transpile/fusion.hpp:43]: It is the second because CNOT-diagonal-CNOT patttern is detected.

I implemented [more efficient version](https://github.com/hhorii/qiskit-aer/blob/9f593798e0a9d9be4bcc7549797d7a3b0349a8fe/src/transpile/fusion_method.hpp#L160-L180) to generate a diagonal matrix but performance improvement was very limited for generation of small diagonal matrix.

File: src/transpile/fusion.hpp
Comment (hhorii) [reply src/transpile/fusion.hpp]: I took it.

**Review by vvilpas — APPROVED** [review]

**Pull request merged**
